### ISSUE TITLE
Error with Windows Core Voices on Windows 11

### ISSUE THREAD

**Issue by andypillip** [nvaccess/nvda#14576]
### Steps to reproduce:

* Install NVDA on a fresh installation of Windows 11
* Launch it

I also went and installed speech modules for Narrator via Windows settings after NVDA didn’t announce anything.

In the narrator options they are separated into Natural voices and Legacy voices, but it’s unclear how to install either. Both buttons send you to the regional settings, which does not at all differentiate between Natural or Legacy. 

### Actual behavior:

* The startup-sound is played
* Nothing is being read

### Expected behavior:

* The NVDA startup dialog should be read out loud

### NVDA logs, crash dumps and other attachments:

[nvda_log.txt](https://github.com/nvaccess/nvda/files/10474289/nvda_log.txt)

Interesting lines:

```
INFO - synthDriverHandler.setSynth (12:30:10.880) - MainThread (1664):
Loaded synthDriver oneCore
ERROR - NVDAHelperLocal (12:30:11.348) - Dummy-4 (5720):
Thread 5720, build\x86\localWin10\oneCoreSpeech.cpp, speak, 345:
Error -2147221164: Class not registered

ERROR - synthDrivers.oneCore.OneCoreSynthDriver._handleSpeechFailure (12:30:11.348) - Dummy-4 (5720):
OneCore synthesizer failed to speak
INFO - NVDAHelperLocal (12:30:31.352) - MainThread (1664):
Thread 1664, build\x86\localWin10\oneCoreSpeech.cpp, ocSpeech_terminate, 236:
ocSpeech_terminate, token: 03DA8C7C
```

#### NVDA installed/portable/running from source:

Installed

#### NVDA version:

2022.4

#### Windows version:

Windows 11 Pro N
Version	10.0.22621 Build 22621

#### Name and version of other software in use when reproducing the issue:

Gnome Boxes 43.2

#### Other information about your system:

Windows is running inside a QEMU virtual machine

#### Does the issue still occur after restarting your computer?

Yes

#### Have you tried any other versions of NVDA? If so, please report their behaviors.

No

#### If NVDA add-ons are disabled, is your problem still occurring?

There are no add-ons installed, it’s a fresh copy.

#### Does the issue still occur after you run the COM Registration Fixing Tool in NVDA's tools menu?

Yes

**Comment by seanbudd** [comment]
Can you follow the instructions to "Add a TTS voice to your PC" from [these docs](https://support.microsoft.com/en-us/windows/appendix-a-supported-languages-and-voices-4486e345-7730-53da-fcfe-55cc64300f01).

My assumption is that since Windows is running under an emulated environment, no TTS voices have been installed.
If you cannot install voices in the emulated environment, NVDA will not be able to use them.

**Comment by seanbudd** [comment]
Yes, that should occur after 5 failures @Brian1Gaff

**Comment by andypillip** [comment]
I did install voices later and then it worked. Just out of the box it didn’t read anything.

If this is only the case in emulated environments like QEMU, it’s less of an issue, of course.

Should the fallback be improved or would you close the ticket because it’s in rare, impractical cases only.

**Comment by seanbudd** [comment]
Closing this as won't/can't fix.
I don't know how the fallback cases could improve, but if you have an idea, feel free to open a new issue with a proposal.

**Comment by XLTechie** [comment]
The fact that it can happen at all is concerning.

NVDA should never be left in a no-speech situation when (1) speech has been 
enabled, and (2) sound is available.
IMO There should always be a working fallback.

@seanbudd can you keep this open to keep it on the radar? There have been other 
(now closed) issues about this happening under various circumstances, and 
reports on mailing lists, etc.

It seems clear that there is some room for improvement here, even if nobody is 
yet sure how to get there from here. Just because we don't know what the 
solution is, doesn't mean it isn't a problem.

**Comment by seanbudd** [comment]
@XLTechie how do you suggest we improve handling this?
Currently after 5 attempts fail to read speech consecutively, we revert to eSpeak.

There is a working fallback in this case.

**Comment by XLTechie** [comment]
@seanbudd 
> There is a working fallback in this case.

Perhaps I'm missing something, but in the case of this issue, the fallback definitely didn't work.
At least, according to the text of the issue ("The startup-sound is played
Nothing is being read"), and to what the OP said in https://github.com/nvaccess/nvda/issues/14576#issuecomment-1433540286 ("Just out of the box it didn’t read anything").

Additionally, none of the info level "Falling back to (next|previous) synthDriver" messages are shown in the log given.

I'm therefore not clear on why you say the fallback worked.

**Comment by seanbudd** [comment]
@XLTechie the log snippet only shows one failure of text being read. i.e. if a user moves around 5 times without speech, the fallback starts. We could decrease that interval?

**Comment by XLTechie** [comment]
@seanbudd  asked:
> @XLTechie how do you suggest we improve handling this? Currently after 5 attempts fail to read speech consecutively, we revert to eSpeak.

setSynth logs "`Loaded synthDriver oneCore`", but then OneCore fails to speak. Is it expected behavior, for OneCore to load successfully, but to simultaneously report a "`Class not registered`" error? Shouldn't the latter negate the former? I.e. shouldn't it, in that case, not even try to speak through OneCore, since if the class isn't registered, I would think it means that OneCore isn't provisioned with voices, so we know it isn't going to work when we try to send speech to it. It shouldn't even require a failure of speech to know OneCore is DOA given that, unless I'm wrong about what the class error means.

Would it be reasonable to assume, that if OneCore voices don't exist (which is presumably the only way we get to this state), that SAPI5 voices, at least the Microsoft ones, will also fail? So instead of trying the normal "next available" fallback, it should jump straight to eSpeak, which is almost guaranteed to work?

The log shows that it loaded SAPI5 after that, but nothing about what happened when SAPI5 tried to speak, nor even what caused SAPI5 to load.

**Comment by XLTechie** [comment]
@seanbudd are there cases where OneCore is otherwise fine, but fails to speak for one or two attempts, then resumes speaking?

Put another way: why is it set to five attempts now?

That said, given that the next synth attempted was SAPI5 instead of eSpeak, I don't actually know if the failover failed, or if there is a more obscure problem. I'm no expert on OneCore API behaviors by any means.

All that seems apparent to me, is that
1. the user didn't get speech because OneCore had no voices;
2. setSynth had a CLASS NOT REGISTEREDerror that (I suspect) should have told it (though it's not currently set to look) that OneCore was never going to speak on this environment, so there was no point in letting the five attempts play out; and
3. the process of synth failover needs some more info level logging.

**Comment by seanbudd** [comment]
>setSynth logs "Loaded synthDriver oneCore", but then OneCore fails to speak. Is it expected behavior, for OneCore to load successfully, but to simultaneously report a "Class not registered" error? Shouldn't the latter negate the former? I.e. shouldn't it, in that case, not even try to speak through OneCore, since if the class isn't registered, I would think it means that OneCore isn't provisioned with voices, so we know it isn't going to work when we try to send speech to it. It shouldn't even require a failure of speech to know OneCore is DOA given that, unless I'm wrong about what the class error means.

That first error occurs when oneCore fails to speak. OneCore is registered successfully, but fails to speak as no voices are set up.
Given that OneCore successfully registers, how should NVDA detect that OneCore isn't provisioned with voices? How can NVDA differentiate that from a one-off random error when producing speech?

>Would it be reasonable to assume, that if OneCore voices don't exist (which is presumably the only way we get to this state), that SAPI5 voices, at least the Microsoft ones, will also fail? So instead of trying the normal "next available" fallback, it should jump straight to eSpeak, which is almost guaranteed to work?
>The log shows that it loaded SAPI5 after that, but nothing about what happened when SAPI5 tried to speak, nor even what caused SAPI5 to load.

Information from the author or a debug log would be required to know this. My assumption is the author changed the synthesizer to SAPI5, which would also not work. SAPI5 also doesn't have this fallback code, as it is not the default synthesizer.

>Put another way: why is it set to five attempts now?

Because speech may fail once randomly, and a user may not want it to fallback to eSpeak immediately.

**Comment by seanbudd** [comment]
> 2. setSynth had a CLASS NOT REGISTEREDerror that (I suspect) should have told it (though it's not currently set to look) that OneCore was never going to speak on this environment, so there was no point in letting the five attempts play out; and

This part is not true.

> 3. the process of synth failover needs some more info level logging.

It may be worth raising this from DEBUGWARNING, however this log does not contain a failover event

**Comment by seanbudd** [comment]
Note the error location:
>Thread 5720, build\x86\localWin10\oneCoreSpeech.cpp, speak, 345:

**Comment by XLTechie** [comment]
I see about SAPI5, that probably was user actuated based on the timing info (about a 20 second separation).

**Comment by XLTechie** [comment]
Ah, I misread the thread IDs there, I see the class error was thrown from _handleSpeechFailure.

Is there no other way than by speaking actual words, to determine that OneCore is in a fail state?
Even hackishly, by speaking silence (a character which doesn't produce speech, but will error the same as real speech)?

If there is, perhaps a first failure to speak, could provoke a sequence of four automatic attempts, and if they all fail, result in fallback?

Alternatively, could we not just try to speak the failed chunk again a few times, and if they all fail, know we have a problem?

The base idea being, that the user shouldn't have to try to read five things in order for NVDA to know the synth has failed.

**Comment by seanbudd** [comment]
>Is there no other way than by speaking actual words, to determine that OneCore is in a fail state?
We are not aware of one.

>If there is, perhaps a first failure to speak, could provoke a sequence of four automatic attempts, and if they all fail, result in fallback?
>Alternatively, could we not just try to speak the failed chunk again a few times, and if they all fail, know we have a problem?
I think these were raised and briefly attempted in #12647.
This could be possible but would be quite complex to implement due to how NVDA interacts with the synthesizer through callbacks.

>The base idea being, that the user shouldn't have to try to read five things in order for NVDA to know the synth has failed.

Reducing the interval to 2 might be more appropriate.
